File: java-home-inventory-core/src/br/com/sailboat/homeinventory/core/interactor/SaveProduct.java
package br.com.sailboat.homeinventory.core.interactor;

import br.com.sailboat.homeinventory.core.entity.Product;
import br.com.sailboat.homeinventory.core.interactor.validator.ProductValidator;
import br.com.sailboat.homeinventory.core.repository.ProductRepository;

public class SaveProduct implements UseCase {

    private Product product;
    private ProductRepository productRepository;

    public SaveProduct(Product product, ProductRepository productRepository) {
        this.product = product;
        this.productRepository = productRepository;
    }

    @Override
    public void execute(Response response) {
        try {
            ProductValidator.validate(product);
            productRepository.save(product);
            response.onSuccess();
        } catch (Exception e) {
            response.onFail(e);
        }
    }

}

File: android-home-inventory-app/app/src/main/kotlin/br/com/sailboat/homeinventory/domain/usecase/SaveProduct.kt
package br.com.sailboat.homeinventory.domain.usecase

import br.com.sailboat.homeinventory.domain.None
import br.com.sailboat.homeinventory.domain.entity.EntityHelper
import br.com.sailboat.homeinventory.domain.entity.Product
import br.com.sailboat.homeinventory.domain.repository.ProductRepository
import javax.inject.Inject

class SaveProduct @Inject constructor(private val productRepository: ProductRepository) : UseCase2<Product, None>() {

    override fun execute(product: Product): None {
        if (product.id == EntityHelper.NO_ID) {
            productRepository.insert(product)
        } else {
            productRepository.update(product)
        }

        return None()
    }

}